exporter/sentryexporter/sentry_exporter.go

```diff
@@ -4,511 +4,772 @@
+							projectsBeingCreated++
+							continue
+						}
+						logger.Error("Failed to get endpoint for project on retry",
+							zap.String("project", key.slug),
+							zap.Error(retryErr))
+						errs = errors.Join(errs, retryErr)
+						continue
+					}
+
+					if retryErr := s.sendTracesToEndpoint(ctx, logger, traces, newEndpoint); retryErr != nil {
+						logger.Error("Failed to send traces to project on retry",
+							zap.String("project", key.slug),
+							zap.Error(retryErr))
+						errs = errors.Join(errs, retryErr)
+						continue
+					}
+					continue
+				}
+			}
+
+			logger.Error("Failed to send traces to project",
+				zap.String("project", key.slug),
+				zap.Error(err))
+			errs = errors.Join(errs, err)
+		}
 	}
-}
 
-// sentryEventFromError creates a sentry event from error event in a span
-func sentryEventFromError(errorMessage, errorType string, span *sentry.Span) (*sentry.Event, error) {
-	if errorMessage == "" && errorType == "" {
-		err := errors.New("error type and error message were both empty")
-		return nil, err
+	if projectsBeingCreated > 0 {
+		logger.Warn("Data dropped for projects being created asynchronously, future data will succeed",
+			zap.Int("projects_being_created", projectsBeingCreated))
 	}
-	event := sentry.NewEvent()
-	event.EventID = generateEventID()
-
-	event.Contexts["trace"] = sentry.TraceContext{
-		TraceID:      span.TraceID,
-		SpanID:       span.SpanID,
-		ParentSpanID: span.ParentSpanID,
-		Op:           span.Op,
-		Description:  span.Description,
-		Status:       span.Status,
-	}.Map()
-
-	event.Type = errorType
-	event.Message = errorMessage
-	event.Level = "error"
-	event.Exception = []sentry.Exception{{
-		Value: errorMessage,
-		Type:  errorType,
-	}}
-
-	event.Sdk.Name = otelSentryExporterName
-	event.Sdk.Version = otelSentryExporterVersion
-
-	event.StartTime = span.StartTime
-	event.Tags = span.Tags
-	event.Timestamp = span.EndTime
-	event.Transaction = span.Description
-
-	return event, nil
-}
 
-// classifyAsOrphanSpans iterates through a list of possible orphan spans and tries to associate them
-// with a transaction. As the order of the spans is not guaranteed, we have to recursively call
-// classifyAsOrphanSpans to make sure that we did not leave any spans out of the transaction they belong to.
-func classifyAsOrphanSpans(orphanSpans []*sentry.Span, prevLength int, idMap map[sentry.SpanID]sentry.SpanID, transactionMap map[sentry.SpanID]*sentry.Event) []*sentry.Span {
-	if len(orphanSpans) == 0 || len(orphanSpans) == prevLength {
-		return orphanSpans
+	if errs != nil {
+		return consumererror.NewPermanent(errs)
 	}
+	return nil
+}
 
-	newOrphanSpans := make([]*sentry.Span, 0, prevLength)
+// sendTracesToEndpoint sends traces to a specific endpoint
+func (s *endpointState) sendTracesToEndpoint(ctx context.Context, logger *zap.Logger, td ptrace.Traces, endpoint *otlpEndpoints) error {
+	request := ptraceotlp.NewExportRequestFromTraces(td)
+	data, err := request.MarshalProto()
```

Comment:
I'd experiment with this, so tackle it last, but `request.MarshalProto` requires us to allocate bytes for request. Maybe we can pre-allocate a shared pool of bytes (via `sync.Pool` or similar) so we only have to allocate once. We can just defer the `bytes.Buffer` reset on this method then.